# Matrix and Vector Operations

AIT 525 assignment on the UCI Adult dataset.

## Setup

Before the assignment's matrix operations, we need to load the dataset, give its columns names, and take a quick first look at it.

In [4]:
import numpy as np
import pandas as pd

columns = [
    "age", "workclass", "fnlwgt", "education", "education-num",
    "marital-status", "occupation", "relationship", "race", "sex",
    "capital-gain", "capital-loss", "hours-per-week", "native-country", "income",
]

df = pd.read_csv("adult_data/adult.data", names=columns, skipinitialspace=True)
df.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


Only 6 of the 15 columns are actual numbers. We'll pull those out into a matrix called `X`, then look at simple summary stats for each one — this also covers the "descriptive statistics" deliverable for Part 1.

In [5]:
numeric_columns = ["age", "fnlwgt", "education-num", "capital-gain", "capital-loss", "hours-per-week"]

df[numeric_columns].describe()

,age,fnlwgt,education-num,capital-gain,capital-loss,hours-per-week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [6]:
X = df[numeric_columns].to_numpy()
print(X.shape)
X[:5]

(32561, 6)


array([[    39,  77516,     13,   2174,      0,     40],
       [    50,  83311,     13,      0,      0,     13],
       [    38, 215646,      9,      0,      0,     40],
       [    53, 234721,      7,      0,      0,     40],
       [    28, 338409,     13,      0,      0,     40]])

## Part 1: Basic Matrix and Vector Operations

### 1. Matrix Addition

We'll take two small batches of rows from `X` and add them together. Adding two batches is like combining two groups of people into one running total.

In [7]:
batch_a = X[:5]
batch_b = X[5:10]

batch_sum = batch_a + batch_b
print(batch_sum)

[[    76 362098     27   2174      0     80]
 [    99 243498     18      0      0     29]
 [    90 425288     18      0      0     85]
 [    84 280502     21  14084      0     90]
 [    70 497858     26   5178      0     80]]


### 2. Matrix Subtraction

Subtracting shows the difference between two batches — useful for comparing groups (e.g., "how does this batch differ from that one?").

In [8]:
batch_diff = batch_a - batch_b
print(batch_diff)

[[      2 -207066      -1    2174       0       0]
 [      1  -76876       8       0       0      -3]
 [    -14    6004       0       0       0      -5]
 [     22  188940      -7  -14084       0     -10]
 [    -14  178960       0   -5178       0       0]]


### 3. Matrix Multiplication

- **Multiplying by an identity matrix** — a special matrix that changes nothing — proves multiplication works as expected.
- **Multiplying by a scaling matrix** shrinks or grows each column by a chosen amount. This is exactly what "feature scaling" means in AI: making columns comparable so no single one dominates a model.

In [9]:
identity = np.eye(6)
batch_a_identity = batch_a @ identity
print(batch_a_identity)
print('\n')

scaling = np.diag([2, 1, 1, 1, 1,  1])
batch_a_scale = batch_a_identity @ scaling
print(batch_a_scale)

[[3.90000e+01 7.75160e+04 1.30000e+01 2.17400e+03 0.00000e+00 4.00000e+01]
 [5.00000e+01 8.33110e+04 1.30000e+01 0.00000e+00 0.00000e+00 1.30000e+01]
 [3.80000e+01 2.15646e+05 9.00000e+00 0.00000e+00 0.00000e+00 4.00000e+01]
 [5.30000e+01 2.34721e+05 7.00000e+00 0.00000e+00 0.00000e+00 4.00000e+01]
 [2.80000e+01 3.38409e+05 1.30000e+01 0.00000e+00 0.00000e+00 4.00000e+01]]


[[7.80000e+01 7.75160e+04 1.30000e+01 2.17400e+03 0.00000e+00 4.00000e+01]
 [1.00000e+02 8.33110e+04 1.30000e+01 0.00000e+00 0.00000e+00 1.30000e+01]
 [7.60000e+01 2.15646e+05 9.00000e+00 0.00000e+00 0.00000e+00 4.00000e+01]
 [1.06000e+02 2.34721e+05 7.00000e+00 0.00000e+00 0.00000e+00 4.00000e+01]
 [5.60000e+01 3.38409e+05 1.30000e+01 0.00000e+00 0.00000e+00 4.00000e+01]]


### 4. Dot Product

The dot product combines two lists of numbers (vectors) into a single number. It's how AI models measure how "similar" or "aligned" two things are.

In [10]:
batch_dot = np.dot(X[:, 0], X[:, 5])
print(batch_dot)

51176886


**Deliverables:**
- Descriptive statistics of the dataset (done in Step 1 above).
- Code demonstrating addition, subtraction, multiplication, and dot product.
- Explanation of each operation and its purpose.
- Discussion of how these operations apply in AI (data transformation, feature scaling, projections).

**Answers:**

*Descriptive statistics* — computed in Setup with `.describe()`: average age is ~38.6 years, average hours worked per week is ~40.4, and `capital-gain` is heavily skewed (mostly 0, but ranges up to 99,999).

*What each operation does and why:*
- **Addition** (`batch_sum`) combined two 5-row batches into one running total per column — the same mechanic used to merge or aggregate separate data batches (Lipschutz & Lipson, 2018).
- **Subtraction** (`batch_diff`) captured how each person in `batch_a` differs from the matching person in `batch_b`, feature by feature — useful for directly comparing two groups or records (Lipschutz & Lipson, 2018).
- **Multiplication** confirmed multiplying by the identity matrix leaves data unchanged, while multiplying by our scaling matrix doubled only the `age` column — showing how a matrix can selectively rescale individual features (Lipschutz & Lipson, 2018; GeeksforGeeks, 2025).
- **Dot product** collapsed the entire `age` and `hours-per-week` columns into a single number (51,176,886), summarizing how those two features move together across all 32,561 people (Lipschutz & Lipson, 2018).

*Why this matters in AI:*
- **Data transformation** — combining and comparing batches (addition/subtraction) is the same basic operation used when merging mini-batches or measuring how a dataset shifts over time.
- **Feature scaling** — multiplying by a diagonal scaling matrix is literally how features get rescaled (e.g., normalization) before training, so no single large-range feature (like `fnlwgt`) dominates a model over a small-range one (like `age`).
- **Projections** — the dot product is the core operation behind similarity measures (like cosine similarity) and is exactly what a single neuron in a neural network computes: a weighted sum of its inputs.

## Part 2: Matrix Transpose and Inverses

### 1. Matrix Transpose

Transpose just flips a matrix — rows become columns. It's used constantly inside AI training (like gradient descent).

In [24]:
print(f'Matrix: \n{X}')
print('\n')

trans_X = X.T
print(f'Matrix Transpose: \n{trans_X}')

Matrix: 
[[    39  77516     13   2174      0     40]
 [    50  83311     13      0      0     13]
 [    38 215646      9      0      0     40]
 ...
 [    58 151910      9      0      0     40]
 [    22 201490      9      0      0     20]
 [    52 287927      9  15024      0     40]]


Matrix Transpose: 
[[    39     50     38 ...     58     22     52]
 [ 77516  83311 215646 ... 151910 201490 287927]
 [    13     13      9 ...      9      9      9]
 [  2174      0      0 ...      0      0  15024]
 [     0      0      0 ...      0      0      0]
 [    40     13     40 ...     40     20     40]]


### 2. Matrix Inverse

`X` itself isn't square, so it can't be inverted directly. We'll build `XᵀX` — multiplying `X` by its own transpose always gives a square matrix. This exact matrix (`XᵀX`) is the key ingredient in the "normal equation," the formula that solves linear regression directly.

Inverse is the "undo" of a matrix — like how dividing undoes multiplying for regular numbers. We'll invert `XᵀX` to see why that's needed to solve for regression weights.

In [28]:
inverse_X = np.linalg.inv(trans_X @ X)
print(f'Matrix Inverse: \n{inverse_X}')

Matrix Inverse: 
[[ 1.32897921e-07 -2.04190395e-12 -2.57265321e-07 -5.44426303e-12
  -1.50906002e-10 -4.82914439e-08]
 [-2.04190395e-12  2.39872990e-15 -2.09781573e-11  1.33440945e-15
   1.83464350e-14 -3.70738390e-12]
 [-2.57265321e-07 -2.09781573e-11  3.11638984e-06 -7.00160712e-11
  -1.17792853e-09 -3.98382019e-07]
 [-5.44426303e-12  1.33440945e-15 -7.00160712e-11  5.69540781e-13
   4.29325429e-13 -1.84775886e-12]
 [-1.50906002e-10  1.83464350e-14 -1.17792853e-09  4.29325429e-13
   1.90927256e-10 -9.21940231e-11]
 [-4.82914439e-08 -3.70738390e-12 -3.98382019e-07 -1.84775886e-12
  -9.21940231e-11  1.67643483e-07]]


**Deliverables:**
- Code showing the matrix transpose and inverse computations.
- Discussion of why these operations are important in AI, particularly in optimization and equation solving.

**Answers:**

*What each operation does and why:*
- **Transpose** (`trans_X`) flips `X`'s rows into columns with `X.T`. We needed this because `X` is 32561×6 — not square — so it can't be inverted directly. Multiplying `X` by its transpose (`XᵀX`) gives us a square 6×6 matrix, which *can* be inverted (Lipschutz & Lipson, 2018).
- **Inverse** (`inverse_X`) is the "undo" of multiplication, the same way dividing undoes multiplying for regular numbers (`A × A⁻¹` gives the identity matrix, just like `5 × 2 ÷ 2` gets you back to 5) (Lipschutz & Lipson, 2018). We computed `np.linalg.inv(trans_X @ X)` to invert that square `XᵀX` matrix.

*Why this matters in AI:*
- **Equation solving** — `(XᵀX)⁻¹` is the key piece of the "normal equation," `weights = (XᵀX)⁻¹ Xᵀy`, which solves linear regression directly in one step (Rodriguez & Zieffler, 2025).
- **Optimization** — without being able to invert `XᵀX`, there'd be no direct formula for the best-fit weights. We'd have to fall back on slower, iterative methods like gradient descent instead of solving for the answer immediately.

## Part 3: Matrix Decomposition and Rank

### 1. Matrix Rank

Rank is a number that tells you how much real, independent information is in a matrix. If a matrix has 6 columns but a rank of only 4, it means 2 of those columns aren't adding anything new — they can be built from the others. In AI, low rank is a warning sign: some features may be redundant.

In [29]:
rank = np.linalg.matrix_rank(X)
print(f'Matrix Rank: \n{rank}')

Matrix Rank: 
6


**Deliverables:**
- Code that calculates the matrix rank.
- Discussion of its implications for data analysis in AI.

**Answers:**

Rank tells you how many columns actually carry unique information (Lipschutz & Lipson, 2018). Since `X`'s rank is 6 (same as its column count), all 6 features are independent — no redundancy. If the rank had been lower, say 4, it would mean 2 columns could be built from the others, and those columns could be dropped without losing information — reducing redundancy, speeding up training, and preventing the model from wasting effort on repeated information (Rodriguez & Zieffler, 2025).

## Part 4: Vector Spaces and Linear Independence

### 1. Linear Independence Test

A set of columns is linearly independent if none of them can be written as a combination of the others — each one brings genuinely new information. We'll check this using rank again: if the rank equals the number of columns we're testing, they're independent. This matters in AI because feeding a model redundant features wastes effort and can confuse training.

In [34]:
is_independent = X.shape[1] == rank
print(f'Number of columns: {X.shape[1]}')
print(f'Rank: {rank}')
print(f'Linearly independent: {is_independent}')

Number of columns: 6
Rank: 6
Linearly independent: True


**Deliverables:**
- Code showing how you test for linear independence.
- Discussion of the importance of linear independence in feature selection for AI models.

**Answers:**

Linear independence matters for feature selection because redundant columns don't add new information — they just waste computation and can confuse a model during training without improving its accuracy (Lipschutz & Lipson, 2018; Rodriguez & Zieffler, 2025). In our case, since `X`'s columns are all independent (rank = number of columns = 6), we know all 6 features are worth keeping — none of them need to be removed.

## References

Becker, B., & Kohavi, R. (1996). *Adult* [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C5XW20

GeeksforGeeks. (2025, July 11). *Different operations on matrices*. https://www.geeksforgeeks.org/engineering-mathematics/different-operation-matrices/

Lipschutz, S., & Lipson, M. L. (2018). *Schaum's outline of linear algebra* (6th ed.). McGraw-Hill Education.

Rodriguez, M., & Zieffler, A. (2025). *Matrix algebra for educational scientists*. https://zief0002.github.io/matrix-algebra/index.html

Tutorialspoint. (n.d.). *NumPy - Matrix library*. Retrieved September 16, 2026, from https://www.tutorialspoint.com/numpy/numpy_matrix_library.htm